File: ldif/ldif-modules/ldif-r2r/ldif-r2r-common/src/main/java/de/fuberlin/wiwiss/r2r/parser/Node.java
/* 
 * LDIF
 *
 * Copyright 2011-2013 Freie Universität Berlin, MediaEvent Services GmbH & Co. KG
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package de.fuberlin.wiwiss.r2r.parser;

/**
 * Created by IntelliJ IDEA.
 * User: andreas
 * Date: 03.05.11
 * Time: 18:55
 * To change this template use File | Settings | File Templates.
 */
public class Node {
	private String value, datatypeOrLanguage;
	private NodeType nodeType;

  Node(String value, String datatypeOrLanguage, NodeType nodeType) {
	  this.value = value;
	  this.datatypeOrLanguage = datatypeOrLanguage;
	  this.nodeType = nodeType;
  }

  public String datatype() {
  	if(nodeType == NodeType.TYPEDLITERAL)
  		return datatypeOrLanguage;
  	else
  		return null;
  }

  public String language() {
  	if(nodeType == NodeType.LANGUAGELITERAL)
  		return datatypeOrLanguage;
  	else
  		return null;
  }

  public String value() {
  	return value;
  }

  public NodeType nodeType() {
  	return nodeType;
  }

  public static Node createLiteral(String value) {
  	return new Node(value, null, NodeType.LITERAL);
  }

  public static Node createTypedLiteral(String value, String dataType) {
  	return new Node(value,dataType, NodeType.TYPEDLITERAL);
  }

  public static Node createLanguageLiteral(String value, String language) {
  	return new Node(value, language, NodeType.LANGUAGELITERAL);
  }

  public static Node createBlankNode(String value) {
  	return new Node(value, null, NodeType.BLANKNODE);
  }

  public static Node createUriNode(String value) {
  	return new Node(value, null, NodeType.URINODE);
  }

  public static Node createVariableNode(String value) {
  	return new Node(value, null, NodeType.VARIABLENODE);
  }

  public static Node createConstantLiteral(String value) {
  	return new Node(value, null, NodeType.CONSTANTLITERAL);
  }

  public static Node createUnknownLexicalNode(String value) {
      return new Node(value, null, NodeType.UNKNOWNLEXICALVALUE);
  }

  public boolean equals(Object other) {
  	if(this==other)	return true;
  	if(!(other instanceof Node)) return false;
  	Node otherNode = (Node) other;
    return (otherNode.nodeType==nodeType) &&
           compareDTorLang(this.datatypeOrLanguage,otherNode.datatypeOrLanguage) &&
           (this.value.equals(otherNode.value));
  }

  private boolean compareDTorLang(String v1, String v2) {
    if(v1==null)
      return v2==null;
    else
      return v1.equals(v2);
  }

  public int hashCode() {
  	int hash = 1;
  	hash = hash*31 + value.hashCode();
  	hash = hash*31 + (datatypeOrLanguage==null ? 0 : datatypeOrLanguage.hashCode());
  	hash = hash*31 + nodeType.hashCode();
  	return hash;
  }

  public String toString() {
  	StringBuilder sb = new StringBuilder();
  	if(nodeType==NodeType.URINODE)
  		sb.append("<");
  	else if(nodeType==NodeType.LITERAL || nodeType==NodeType.TYPEDLITERAL || nodeType==NodeType.LANGUAGELITERAL)
  		sb.append("\"");
  	else if(nodeType==NodeType.VARIABLENODE)
  		sb.append("?");

  	sb.append(value);

  	if(nodeType==NodeType.URINODE)
  		sb.append(">");
  	else if(nodeType==NodeType.LITERAL || nodeType==NodeType.TYPEDLITERAL || nodeType==NodeType.LANGUAGELITERAL)
  		sb.append("\"");

  	if(nodeType==NodeType.LANGUAGELITERAL)
  		sb.append("@").append(datatypeOrLanguage);
  	if(nodeType==NodeType.TYPEDLITERAL)
  		sb.append("^^<").append(datatypeOrLanguage).append(">");
  	return sb.toString();
  }
}

File: ldif/ldif-core/src/main/scala/ldif/entity/Node.scala
/* 
 * LDIF
 *
 * Copyright 2011-2012 Freie Universität Berlin, MediaEvent Services GmbH & Co. KG
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ldif.entity

import org.semanticweb.yars.nx.parser.NxParser
import ldif.util.NTriplesStringConverter
import ldif.util.Consts

final case class Node(value : String, datatypeOrLanguage : String, nodeType : Node.NodeType, graph : String) extends NodeTrait
{
  def modifyGraph(graph: String): Node = {
    Node(this.value, this.datatypeOrLanguage, this.nodeType, graph)
  }

  def compare(other: Node): Int = super.compare(other)
}

object Node
{
  val defaultGraph : String = Consts.DEFAULT_GRAPH

  def createLiteral(value : String, graph : String = null) = Node(value, null, Literal, graph)

  def createTypedLiteral(value : String, datatype : String, graph : String = null) = Node(value, datatype, TypedLiteral, graph)

  def createLanguageLiteral(value : String, language : String, graph : String = null) = Node(value, language, LanguageLiteral, graph)

   def createBlankNode(value : String, graph : String) = Node(NTriplesStringConverter.convertBnodeLabel(value), null, BlankNode, graph)

  def createUriNode(value : String, graph : String = null) = Node(value, null, UriNode, graph)

  def fromString(value : String, graph : String) = {
    val nxNode = NxParser.parseNode(value)
    fromNxNode(nxNode,graph)
  }

  def fromString(value : String) : Node = fromString(value, defaultGraph)

  def fromNxNode(nxNode : org.semanticweb.yars.nx.Node, graph : String) = {
     nxNode match {
      case lit:org.semanticweb.yars.nx.Literal => {
        val dt = lit.getDatatype
        val lang = lit.getLanguageTag
        val value = lit.getData
        if (dt!=null)
          ldif.entity.Node.createTypedLiteral(value,dt.toString,graph)
        else if (lang!=null)
          ldif.entity.Node.createLanguageLiteral(value,lang,graph)
        else ldif.entity.Node.createLiteral(value,graph)
      }
      case bno:org.semanticweb.yars.nx.BNode =>
        ldif.entity.Node.createBlankNode(bno.toString,graph)
      case res:org.semanticweb.yars.nx.Resource =>
        ldif.entity.Node.createUriNode(res.toString,graph)
    }
  }

  def fromNxNode(nxNode : org.semanticweb.yars.nx.Node) : Node = fromNxNode(nxNode,defaultGraph)

  // Build node from an XML node
  def fromXML(xml : scala.xml.Node) : Node = xml match  {
    //TODO Literal language and datatype not supported in M1
    case <Uri>{value @ _*}</Uri> => createUriNode(value.text,defaultGraph)
    case <Literal>{value @ _*}</Literal> => createLiteral(value.text,defaultGraph)
    case <BlankNode>{value @ _*}</BlankNode> => createBlankNode(value.text,defaultGraph)
  }

  sealed trait NodeType {val id: Int}

  case object Literal extends NodeType { val id = 1}

  case object TypedLiteral extends NodeType { val id = 2}

  case object LanguageLiteral extends NodeType { val id = 3}

  case object BlankNode extends NodeType { val id = 4}

  case object UriNode extends NodeType { val id = 5}

  case object NodeTypeMap {
    val map = Map(1 -> Literal, 2 -> TypedLiteral, 3 -> LanguageLiteral, 4 -> BlankNode, 5 -> UriNode)

    def apply(index: Int) = map(index)
  }
}
